### PR TITLE
Fix flash/sage varlen prep under torch.compile with dynamic shapes

### LINKED ISSUE #11957: Flash/Sage varlen does not work with torch.compile
@a-r-r-o-w @DN6 @tolgacangoz 
I got a bit excited about this PR and wanted to give it a go. I love the syntax, both the setter function and the context, great work!

I wanted to also see if it would still compile but got the following error logs:
```
[t+28s648ms]   0%|          | 0/30 [00:00<?, ?it/s]/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/torch/_dynamo/variables/functions.py:1601: UserWarning: Dynamo detected a call to a `functools.lru_cache`-wrapped function. Dynamo ignores the cache wrapper and directly traces the wrapped function. Silent incorrectness is only a *potential* risk, not something we have observed. Enable TORCH_LOGS="+dynamo" for a DEBUG stack trace.
[t+28s650ms]   torch._dynamo.utils.warn_once(msg)
[t+28s666ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0] Graph break from `Tensor.item()`, consider setting:
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     torch._dynamo.config.capture_scalar_outputs = True
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0] or:
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     env TORCHDYNAMO_CAPTURE_SCALAR_OUTPUTS=1
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0] to include these operations in the captured graph.
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0] Graph break: from user code at:
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/diffusers/models/transformers/transformer_flux.py", line 733, in forward
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     encoder_hidden_states, hidden_states = block(
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/diffusers/models/transformers/transformer_flux.py", line 456, in forward
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     attention_outputs = self.attn(
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/diffusers/models/transformers/transformer_flux.py", line 343, in forward
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     return self.processor(self, hidden_states, encoder_hidden_states, attention_mask, image_rotary_emb, **kwargs)
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/diffusers/models/transformers/transformer_flux.py", line 117, in __call__
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     hidden_states = dispatch_attention_fn(
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/diffusers/models/attention_dispatch.py", line 241, in dispatch_attention_fn
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     return backend_fn(**kwargs)
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/diffusers/models/attention_dispatch.py", line 962, in _sage_varlen_attention
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     _prepare_for_flash_attn_or_sage_varlen(
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/diffusers/models/attention_dispatch.py", line 351, in _prepare_for_flash_attn_or_sage_varlen
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     return _prepare_for_flash_attn_or_sage_varlen_without_mask(batch_size, seq_len_q, seq_len_kv, device)
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/torch/_dynamo/polyfills/__init__.py", line 253, in getattr_and_trace
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     return fn(*args[2:], **kwargs)
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/diffusers/models/attention_dispatch.py", line 321, in _prepare_for_flash_attn_or_sage_varlen_without_mask
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]     max_seqlen_q = seqlens_q.max().item()
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]
[t+28s667ms] W0718 20:32:55.420000 1 /inferencesh/cache/gpu/uv/archive-v0/bZ7CJbyNBkaHsGR2jn6Vv/torch/_dynamo/variables/tensor.py:1048] [0/0]
[t+28s669ms]   0%|          | 0/30 [00:00<?, ?it/s]
[t+28s670ms] [ERROR] Traceback (most recent call last):
[t+28s671ms]   File "/server/tasks.py", line 50, in run_task
[t+28s671ms]     output = await result
[t+28s671ms]              ^^^^^^^^^^^^
[t+28s671ms]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/src/inference.py", line 248, in run
[t+28s671ms]     result = self.pipeline(
[t+28s671ms]              ^^^^^^^^^^^^^^
[t+28s671ms]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
[t+28s671ms]     return func(*args, **kwargs)
[t+28s671ms]            ^^^^^^^^^^^^^^^^^^^^^
[t+28s671ms]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/diffusers/pipelines/flux/pipeline_flux_kontext.py", line 1063, in __call__
[t+28s671ms]     noise_pred = self.transformer(
[t+28s671ms]                  ^^^^^^^^^^^^^^^^^
[t+28s671ms]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 411, in __call__
[t+28s671ms]     return super().__call__(*args, **kwargs)
[t+28s671ms]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[t+28s671ms]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
[t+28s671ms]     return self._call_impl(*args, **kwargs)
[t+28s671ms]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[t+28s671ms]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1784, in _call_impl
[t+28s671ms]     return forward_call(*args, **kwargs)
[t+28s671ms]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[t+28s672ms]   File "/inferencesh/apps/gpu/31s80r2z5hw6jskcs081xh8wdb/venv/3.12/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 812, in compile_wrapper
[t+28s672ms]     raise e.with_traceback(None) from e.__cause__  # User compiler error
[t+28s672ms]     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[t+28s672ms] torch._dynamo.exc.Unsupported: Unsupported Tensor.item() call with capture_scalar_outputs=False
[t+28s672ms]   Explanation: Dynamo does not support tracing `Tensor.item()` with config.capture_scalar_outputs=False.
```

compiling with both
```
self.pipeline.transformer.compile_repeated_blocks(fullgraph=True)
```
and
```
self.pipeline.transformer.to(memory_format=torch.channels_last)
self.pipeline.transformer = torch.compile(
  self.pipeline.transformer, mode="max-autotune", fullgraph=True
)
```

yields the same result

after

```
self.pipeline.transformer.set_attention_backend("sage_varlen")
```

_Originally posted by @okaris in https://github.com/huggingface/diffusers/issues/11916#issuecomment-3090651548_

### PR DESCRIPTION
**Pull request by ShivamShrirao**
# What does this PR do?

Fixes #11957.

That report is the `.item()` graph break on the **no-mask** path (Flux passes no attention mask), which
change 2 below removes entirely — the no-mask helper no longer calls `.item()` at all. #11970 was an
earlier attempt at the same issue and was closed unmerged. Masked varlen still keeps one necessary
`.item()` for the data-dependent key lengths, as noted after the two changes.

Makes the flash-attn / sage varlen attention backends usable under `torch.compile` with dynamic shapes. Two changes in the `_prepare_for_flash_attn_or_sage_varlen_*` helpers, with no numerical change:

1. **`cu_seqlens` is now built with `torch.arange` instead of `cumsum(full(...))`.** Inductor rewrites the cumsum-of-a-constant pattern internally, and that rewrite fails when the fill value is a symbolic sequence length. The lengths are uniform here (every sequence in the batch has the same length), so `torch.arange(0, (batch_size + 1) * seq_len, seq_len)` produces identical offsets directly — and is cheaper.

   Minimal repro (fails on torch 2.6.0 and 2.12.1):

   ```python
   import torch

   @torch.compile(dynamic=True, fullgraph=True)
   def f(x):
       b, s = x.shape[0], x.shape[1]
       seqlens = torch.full((b,), s, dtype=torch.int32, device=x.device)
       cu = torch.zeros(b + 1, dtype=torch.int32, device=x.device)
       cu[1:] = torch.cumsum(seqlens, dim=0)
       return cu

   f(torch.empty(2, 77, 8))
   # BackendCompilerFailed: inductor raised:
   # TypeError: unsupported operand type(s) for *: 'FakeTensor' and 'Node'
   ```

2. **`max_seqlen_q`/`max_seqlen_k` are returned as the already-known Python int instead of `seqlens.max().item()`.** The `.item()` forces a GPU→CPU sync and a graph break; the max of uniform lengths is just `seq_len` itself, which the caller already has as an int. Same type as before (`.item()` also returned an int), so all call sites are unaffected.

In the masked helper only the query side changes: key lengths are data-dependent (`attn_mask.sum(dim=1)`), so their cumsum does not hit the inductor rewrite and their `.item()` is inherent to varlen with padding.

Found while enabling flash-attn varlen + `torch.compile` for the Bria Fibo pipelines, but the helpers are shared by every model using these backends.

## Verification

- The new construction is bit-identical to the old one across batch/seq-length combinations, for both the masked and unmasked helpers.
- The repro above fails identically on torch 2.6.0 (cu124) and torch 2.12.1 (cu130).
- With the fix, the real `flash_attn_varlen_func` (flash-attn 2.8.3, H200) compiles with `torch.compile(dynamic=True, fullgraph=True)` across changing batch/sequence shapes and matches eager output exactly (max diff 0.0).

## Before submitting
- [x] Did you use an AI agent (Claude Code, Codex, Cursor, etc.) to help with this PR? If so:
  - [x] Did you read the [Coding with AI agents](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution#coding-with-ai-agents) guide?
  - [x] Did you run the [`self-review`](https://github.com/huggingface/diffusers/blob/main/.ai/skills/self-review/SKILL.md) skill on the diff?
  - [x] Did you share the final self-review notes in the PR description or a comment?
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/diffusers/main/en/conceptual/contribution)?
- [x] Did you read our [philosophy doc](https://huggingface.co/docs/diffusers/main/en/conceptual/philosophy)?

<details>
<summary>Self-review notes</summary>

Two-function diff reviewed against `.ai/review-rules.md`. **Verdict: READY.** No API change: same return tuple structure, same values, and the max-seqlen entries keep the type `.item()` produced (Python int). Checked all 10 call sites in `attention_dispatch.py`, including the hub-kernel forward/backward ops added recently — the masked-path callers already discard the helper's `max_seqlen_q` and use `seq_len_q` locally, consistent with this change; unmasked-path callers consume the returned ints unchanged (`ctx` scalar attributes, never `save_for_backward`). Equivalence verified numerically against the previous cumsum construction, masked and unmasked. Left for the reviewer: no CPU unit test covers these helpers directly; the behavior is exercised through the GPU flash-attn backend tests.

</details>

## Who can review?

@sayakpaul @DN6

### CODE REVIEW

**Comment by sayakpaul** [comment]
Can you provide a minimal reproducer with the diffusers functions included?

**Comment by ShivamShrirao** [comment]
Reproducer with diffusers varlen backends call:

```python
import torch
from diffusers.models.attention_dispatch import _prepare_for_flash_attn_or_sage_varlen

@torch.compile(dynamic=True, fullgraph=True)
def prep(hidden_states):
    batch_size, seq_len, _ = hidden_states.shape
    return _prepare_for_flash_attn_or_sage_varlen(
        batch_size, seq_len, seq_len, attn_mask=None, device=hidden_states.device
    )

prep(torch.randn(2, 77, 64, device="cuda"))
```

On main:

```python
torch._inductor.exc.InductorError: TypeError: unsupported operand type(s) for *: 'FakeTensor' and 'Node'
```

With the fix:

```python
((tensor([77, 77], device='cuda:0', dtype=torch.int32),
  tensor([77, 77], device='cuda:0', dtype=torch.int32)),
 (tensor([  0,  77, 154], device='cuda:0', dtype=torch.int32),
  tensor([  0,  77, 154], device='cuda:0', dtype=torch.int32)),
 (77, 77))
```
 Tested with torch 2.12.1 and 2.13.0+cu129.

**Comment by sayakpaul** [comment]
Sorry, but this is bit vague to determine which attention backend is causing this, though.

File: src/diffusers/models/attention_dispatch.py [hunk 1]
```diff
@@ -615,13 +614,13 @@ def _prepare_for_flash_attn_or_sage_varlen_with_mask(
...
+    # Queries are uniform, so arange (see the no-mask helper: cumsum(full(...)) breaks inductor
+    # under dynamic shapes). Keys are data-dependent and keep the cumsum.
+    cu_seqlens_q = torch.arange(0, (batch_size + 1) * seq_len_q, seq_len_q, dtype=torch.int32, device=device)
     cu_seqlens_k = torch.zeros(batch_size + 1, dtype=torch.int32, device=device)
-    cu_seqlens_q[1:] = torch.cumsum(seqlens_q, dim=0)
     cu_seqlens_k[1:] = torch.cumsum(seqlens_k, dim=0)
-    max_seqlen_q = seqlens_q.max().item()
     max_seqlen_k = seqlens_k.max().item()
-    return (seqlens_q, seqlens_k), (cu_seqlens_q, cu_seqlens_k), (max_seqlen_q, max_seqlen_k)
+    return (seqlens_q, seqlens_k), (cu_seqlens_q, cu_seqlens_k), (seq_len_q, max_seqlen_k)
```
Comment (sayakpaul) [inline src/diffusers/models/attention_dispatch.py:624]: Why not return `max_seqlen_q` like other?

Cc: @zhtmike do you have any comments here?

Comment (zhtmike) [reply src/diffusers/models/attention_dispatch.py:624]: I think using `seq_len_q` is fine , since query is always a full sequence.

But according to the PR description, the mask path is still broken under torch compile?

Plus, I think we need a test to guard this

**Comment by ShivamShrirao** [comment]
The helper is shared by every varlen backend, so it's all of them rather than one:

- `_flash_varlen_attention`
- `_flash_varlen_attention_3`
- `_sage_varlen_attention`
- `_flash_varlen_attention_hub` and the hub forward ops `_flash_varlen_attention_hub_forward_op` / `_flash_attention_3_varlen_hub_forward_op`

All of them go through `_prepare_for_flash_attn_or_sage_varlen` to build `cu_seqlens`, so any of them fails the moment you wrap it in `torch.compile(dynamic=True, fullgraph=True)`.

File: src/diffusers/models/attention_dispatch.py (hunk 1 above)
Comment (ShivamShrirao) [reply src/diffusers/models/attention_dispatch.py:624]: Yes. queries are always full length, so `max_seqlen_q` is `seq_len_q`, the caller already has it as an int.

The masked path compiles too and is not broken. The remaining .item() is on the key side only, and it doesn't block fullgraph=True. On torch 2.13.0+cu129 with this PR:

```
no-mask: PASS cu_q=[0, 77, 154] cu_k=[0, 77, 154] max=(77, 77)
masked:  PASS cu_q=[0, 77, 154] cu_k=[0, 77, 137] max=(77, 77)
```

The description only meant that the key-side .item() stays, since padded lengths are data-dependent.

**Comment by ShivamShrirao** [comment]
Hi @sayakpaul , any update on this ? How can I help to get this merged ?

**Comment by ShivamShrirao** [comment]
Hi @sayakpaul bumping this up!

File: src/diffusers/models/attention_dispatch.py [hunk 2]
```diff
@@ -598,13 +598,12 @@ def _prepare_for_flash_attn_or_sage_varlen_without_mask(
...
-    cu_seqlens_k = torch.zeros(batch_size + 1, dtype=torch.int32, device=device)
-    cu_seqlens_q[1:] = torch.cumsum(seqlens_q, dim=0)
-    cu_seqlens_k[1:] = torch.cumsum(seqlens_k, dim=0)
-    max_seqlen_q = seqlens_q.max().item()
-    max_seqlen_k = seqlens_k.max().item()
-    return (seqlens_q, seqlens_k), (cu_seqlens_q, cu_seqlens_k), (max_seqlen_q, max_seqlen_k)
+    # Built with arange instead of cumsum(full(...)): inductor rewrites that pattern into
+    # `arange * fill_value`, which raises under dynamic shapes because the fill value is a
+    # symbolic sequence length. The lengths are uniform here, so arange is also cheaper.
+    cu_seqlens_q = torch.arange(0, (batch_size + 1) * seq_len_q, seq_len_q, dtype=torch.int32, device=device)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/models/attention_dispatch.py]: @ShivamShrirao seems like this was not addressed?

[the remark this answers is not in the document]
Comment (ShivamShrirao) [reply src/diffusers/models/attention_dispatch.py]: @sayakpaul Thanks, fixed in 0ac4fbf. The sequence length now scales a unit-step arange instead of being the step itself, so seq_len == 0 returns all-zero offsets like the cumsum path it replaced, instead of raising. Verified it still matches cumsum(full(...)) across batch/seq combinations and still compiles under dynamic=True.

File: src/diffusers/models/attention_dispatch.py [hunk 3]
```diff
@@ -599,13 +599,13 @@ def _prepare_for_flash_attn_or_sage_varlen_without_mask(
...
-    max_seqlen_q = seqlens_q.max().item()
-    max_seqlen_k = seqlens_k.max().item()
-    return (seqlens_q, seqlens_k), (cu_seqlens_q, cu_seqlens_k), (max_seqlen_q, max_seqlen_k)
+    # Built with arange instead of cumsum(full(...)): inductor rewrites that pattern into
+    # `arange * fill_value`, which raises under dynamic shapes because the fill value is a
+    # symbolic sequence length. The lengths are uniform here, so arange is also cheaper.
+    offsets = torch.arange(0, batch_size + 1, dtype=torch.int32, device=device)
+    cu_seqlens_q = offsets * seq_len_q
+    cu_seqlens_k = offsets * seq_len_kv
+    return (seqlens_q, seqlens_k), (cu_seqlens_q, cu_seqlens_k), (seq_len_q, seq_len_kv)
```
Comment (sayakpaul) [inline src/diffusers/models/attention_dispatch.py:608]: Why are we changing the return signature? The caller can reuse the `seq_len_q` and `seq_len_kv` right?

**Review by sayakpaul — COMMENTED** [review]
Thanks! Sorry about the delay. Can we also add some lightweight tests, keeping the convention similar to `tests/models/testing_utils/attention.py`?

File: src/diffusers/models/attention_dispatch.py (hunk 3 above)
Comment (ShivamShrirao) [reply src/diffusers/models/attention_dispatch.py:608]: The shape is unchanged, still three pairs. The third pair was `seqlens.max().item()`, which for uniform lengths is the input by construction, so this just skips a device sync per call. It also matters for compile: `full((b,), n, dtype=int32).max()` with a symbolic `n` fails inductor codegen under `dynamic=True` on `torch 2.12 `(int32/int64 loop-carried type mismatch in the reduction kernel).

**Comment by ShivamShrirao** [comment]
No worries! Thanks for getting back. I have added `TestVarlenAttentionCompile` in `tests/models/test_attention_dispatch.py`: `dispatch_attention_fn` under `flash_varlen_hub` / `_flash_3_varlen_hub`, compiled with `dynamic=True, fullgraph=True` across changing sequence lengths with error_on_recompile=True, compared to eager. It fails on main with the original `FakeTensor * Node` error and passes with this PR. Plus a small zero-length check for the edge case raised above.

File: tests/models/test_attention_dispatch.py [hunk 4]
```diff
@@ -170,3 +173,57 @@ def test_attn_backend_backward_parity(self, cp_type, attention_backend):
...
+            try:
+                with (
+                    torch._inductor.utils.fresh_inductor_cache(),
+                    torch._dynamo.config.patch(error_on_recompile=True),
+                ):
+                    for seq_len in (128, 256, 256):
+                        q, k, v = make_qkv(seq_len)
+                        torch.testing.assert_close(
+                            compiled_attention(q, k, v), dispatch_attention_fn(q, k, v), atol=1e-3, rtol=1e-3
+                        )
```
Comment (sayakpaul) [inline tests/models/test_attention_dispatch.py:220]: Perfect!

File: tests/models/test_attention_dispatch.py [hunk 5]
```diff
@@ -170,3 +173,57 @@ def test_attn_backend_backward_parity(self, cp_type, attention_backend):
...
+            finally:
+                torch.compiler.reset()
+
+    def test_zero_length_sequence(self):
+        _, (cu_seqlens_q, cu_seqlens_k), max_seqlens = _prepare_for_flash_attn_or_sage_varlen(
+            2, 0, 0, device=torch_device
+        )
+        assert cu_seqlens_q.tolist() == cu_seqlens_k.tolist() == [0, 0, 0]
+        assert max_seqlens == (0, 0)
```
Comment (sayakpaul) [inline tests/models/test_attention_dispatch.py]: We can remove this test because it tests a private function.

**Review by sayakpaul — APPROVED** [review]
Very nice work!

File: tests/models/test_attention_dispatch.py (hunk 5 above)
Comment (ShivamShrirao) [reply tests/models/test_attention_dispatch.py]: Removed, thanks!
